### DISCUSSION TITLE
LifecycleManager hydrationMiddleware can't be overridden

### DISCUSSION THREAD

**Discussion by lucernae** [livewire/livewire#3858] [3. Feature ideas]
Hi,

I was trying to figure out how to add my own hydrationMiddleware for livewire. I'm browsing the source code and found out that the middlewares were registered in LivewireServiceProvider class:

https://github.com/livewire/livewire/blob/95dfe608394b49c5076d34f42d1aed3e3de92987/src/LivewireServiceProvider.php#L325-L374

Then, I proceed to override the Service Provider by creating my own. I added my own middleware to the list of hydrationMiddleware. It doesn't work.

After stepping through the debugger, I found out that the registration method in the LifecycleManager class is like this:

https://github.com/livewire/livewire/blob/95dfe608394b49c5076d34f42d1aed3e3de92987/src/LifecycleManager.php#L64-L67

I was wondering if this is just a typo? The += operator is a union assignment. If it is operated on two array, it only compare the keys and not the values. Since the keys are array indices (numbers), it is not possible to change the values itself. If `$classes` contains more keys, only the extra keys got appended. If I inserted my middleware in the middle index, it won't be stored.

# Proposal to fix

If this is indeed just a typo, I propose the following fix (one from any):

- We can change the += operator into an assignment. So that we provide the full order of the middleware when developers want to override it. Then create getters to retrieve the static:: variables.
- Use service container pattern, just like LivewireManager. This way, we can swap out the LifecycleManager with other singleton that we extend ourselves. Then we register it to override the current singleton in developer's Service Provider.

**Comment by joshhanley** [comment]
@lucernae the code you are looking at isn't a typo, but it was written only for internal use, where it's never intended for other middleware to be added or existing middleware to be replaced.

As such I wouldn't call it a bug and I don't think changing it is a good idea at this stage.

But I do think there should be a way we can register our own custom middleware.

The question with that, is how.

Because just appending a middleware isn't exactly ideal as you may actually want to specify exactly where the middleware goes in the stack.

Just to get some more information on this, what middleware were you hoping to add to the stack and where do want it added?

I've change the category from bug to feature request.

Hope this helps!

**Reply by lucernae** [reply]
Thank you @joshhanley . Yes, that is why I asked it first.

My use case is that I need to insert my own hydration/dehydration middleware in the stack, in my own defined order, because for some cases the dehydration order might not be exactly the reverse of hydration.

Currently I solved this by inheriting LifecycleManager class as my OverridesLifecycleManager class. I clear out static::$hydrationMiddleware to an empty array and re-register the hydration middleware and insert my own middleware in the order that I wanted.

> Just to get some more information on this, what middleware were you hoping to add to the stack and where do want it added?

I've made a helper components to generate form fields for a custom widget. I like how model bind work in Livewire, but it's not flexible enough for my use case if the attributes needs to be transformed before it is used by the component. 

In the hydration process, I added a middleware just before PerformDataBindingUpdates. Because public properties and Hydration hook calls needs to be run first, and then I can use the public properties to transform the data into the format that the component likes.

For the dehydration process, I added a middleware after HydratePublicProperties (because public properties and models needs to be dehydrated first, and then I dehydrated some custom fields into the dehydrated properties).

It looks like this:

```php
class LivewireFormFieldServiceProvider extends LivewireServiceProvider
{
    protected function registerHydrationMiddleware()
    {
        OverridesLifecycleManager::clearHydrationMiddleware();
        LifecycleManager::registerHydrationMiddleware([

            /* This is the core middleware stack of Livewire. It's important */
            /* to understand that the request goes through each class by the */
            /* order it is listed in this array, and is reversed on response */
            /*                                                               */
            /* ↓    Incoming Request                  Outgoing Response    ↑ */
            /* ↓                                                           ↑ */
            /* ↓    Secure Stuff                                           ↑ */
            /* ↓ */ SecureHydrationWithChecksum::class, /* --------------- ↑ */
            /* ↓ */ NormalizeServerMemoSansDataForJavaScript::class, /* -- ↑ */
            /* ↓ */ HashDataPropertiesForDirtyDetection::class, /* ------- ↑ */
            /* ↓                                                           ↑ */
            /* ↓    Hydrate Stuff                                          ↑ */
            /* ↓ */ FormFieldDehydrationHooks::class, /* ----------------- ↑ */
            /* ↓ */ HydratePublicProperties::class, /* ------------------- ↑ */
            /* ↓ */ CallPropertyHydrationHooks::class, /* ---------------- ↑ */
            /* ↓ */ CallHydrationHooks::class, /* ------------------------ ↑ */
            /* ↓ */ FormFieldHydrationHooks::class, /* ------------------- ↑ */
            /* ↓                                                           ↑ */
            /* ↓    Update Stuff                                           ↑ */
            /* ↓ */ PerformDataBindingUpdates::class, /* ----------------- ↑ */
            /* ↓ */ PerformActionCalls::class, /* ------------------------ ↑ */
            /* ↓ */ PerformEventEmissions::class, /* --------------------- ↑ */
            /* ↓                                                           ↑ */
            /* ↓    Output Stuff                                           ↑ */
            /* ↓ */ RenderView::class, /* -------------------------------- ↑ */
            /* ↓ */ NormalizeComponentPropertiesForJavaScript::class, /* - ↑ */

        ]);

        LifecycleManager::registerInitialDehydrationMiddleware([

            /* Initial Response */
            /* ↑ */ [SecureHydrationWithChecksum::class, 'dehydrate'],
            /* ↑ */ [NormalizeServerMemoSansDataForJavaScript::class, 'dehydrate'],
            /* ↑ */ [FormFieldDehydrationHooks::class, 'dehydrate'],
            /* ↑ */ [HydratePublicProperties::class, 'dehydrate'],
            /* ↑ */ [CallPropertyHydrationHooks::class, 'dehydrate'],
            /* ↑ */ [CallHydrationHooks::class, 'initialDehydrate'],
            /* ↑ */ [RenderView::class, 'dehydrate'],
            /* ↑ */ [NormalizeComponentPropertiesForJavaScript::class, 'dehydrate'],

        ]);

        LifecycleManager::registerInitialHydrationMiddleware([

            [CallHydrationHooks::class, 'initialHydrate'],
            [FormFieldHydrationHooks::class, 'hydrate'],

        ]);
    }
}
```

Note that for `FormFieldDehydrationHooks::class`, I only implement `dehydrate` method. While on `FormFieldHydrationHooks::class`, I only implement `hydrate` method.

Re-registration works in the current Livewire code because of line `OverridesLifecycleManager::clearHydrationMiddleware();` where I emptied the `static::$hydrationMiddleware` array.
